distsql: streams infrastructure and multi-node test

RaduBerinde opened the pull request:
This change implements streams and adds various pieces of infrastructure.

I split the change into a few commits to help with reviewing (but note that some of them probably don't build separately). Tracked under #7583.

@andreimatei

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/7959)

<!-- Reviewable:end -->

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> (but you're going to want to wait for @andreimatei).

---

Reviewed 6 of 6 files at r1, 7 of 7 files at r2, 2 of 2 files at r3, 2 of 2 files at r4, 4 of 4 files at r5, 2 of 2 files at r6.
Review status: all files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[sql/distsql/flow.go, line 64 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7959#-KNc3pDznsKPFBsEgfea:-KNc3pE-UFkY-lRKD0N5:-235635886) ([raw file](https://github.com/cockroachdb/cockroach/blob/f254bcc6fa60338eb186841acc8ac2f531424a92/sql/distsql/flow.go#L64)):*
> ```Go
> 
> func newFlow(
> 	ctx FlowCtx,
> ```

I wonder whether we should reserve `ctx` for `context.Context` (much like how `err` is reserved for `error`).

---

*[sql/distsql/flow.go, line 263 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7959#-KNc9-djrkhW9fgY3_rQ:-KNc9-djrkhW9fgY3_rR:18861378) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a11041412870fd650d60a25125eb0c48bb02ad1/sql/distsql/flow.go#L263)):*
> ```Go
> 
> func (f *Flow) getInboundStream(sid StreamID) (RowReceiver, error) {
> 	// TODO(radu): detect if we connect to a stream multiple times
> ```

nit: s/$/./

---

*[sql/distsql/flow.go, line 295 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7959#-KNc98QQq1jjjLmeeSUo:-KNc98QQq1jjjLmeeSUp:-605574972) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a11041412870fd650d60a25125eb0c48bb02ad1/sql/distsql/flow.go#L295)):*
> ```Go
> 		f.flowRegistry.UnregisterFlow(f.id)
> 	}
> 	f.status = FlowFinished
> ```

Check against double-`Cleanup()`?

---

*[sql/distsql/inbound.go, line 60 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7959#-KNc6h1CQJgCOlo371bv:-KNc6h1CQJgCOlo371bw:-1895510741) ([raw file](https://github.com/cockroachdb/cockroach/blob/2dc8311c371f3fe95bf18fd07e75e56dfdc6bf50/sql/distsql/inbound.go#L60)):*
> ```Go
> 			row, err := sd.GetRow(nil)
> 			if err != nil {
> 				retErr = err
> ```

Instead of all of these `retErr = err; break loop`, could you `return responseError(err)`  which is roughly

```go
responseError := func(err error) error {
  dst.Close(err)
  resp := &SimpleResponse{}
  resp.Error = something(err)
  return stream.SendAndClose(resp)
}
```

which would obviate I think all of the `break loop`?

---

*[sql/distsql/joinreader.go, line 97 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7959#-KNc46yRlmISFMXAYhAR:-KNc46ySNRLesJI4UOkr:699938301) ([raw file](https://github.com/cockroachdb/cockroach/blob/f254bcc6fa60338eb186841acc8ac2f531424a92/sql/distsql/joinreader.go#L97)):*
> ```Go
> 
> 	if log.V(2) {
> 		// TODO(radu): add info about the joinreader in the context.
> ```

The TODO (or at least its solution) should be in the constructor.

---

*[sql/distsql/outbox.go, line 64 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7959#-KNc4wJdnJNYL8LOJ1_M:-KNc4wJdnJNYL8LOJ1_N:1860073472) ([raw file](https://github.com/cockroachdb/cockroach/blob/be5a1469549b4526af1333981045f63df2559e06/sql/distsql/outbox.go#L64)):*
> ```Go
> var _ RowReceiver = &outbox{}
> 
> func newOutboxSimpleFlowStream(logCtx context.Context, stream DistSQL_RunSimpleFlowServer) *outbox {
> ```

Seems weird to name a `context.Context` not `ctx`.

---

*[sql/distsql/outbox.go, line 172 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7959#-KNc5hrfeeKq8wFCBxQk:-KNc5hrfeeKq8wFCBxQl:597904721) ([raw file](https://github.com/cockroachdb/cockroach/blob/be5a1469549b4526af1333981045f63df2559e06/sql/distsql/outbox.go#L172)):*
> ```Go
> 	err := m.mainLoop()
> 
> 	m.flushTicker.Stop()
> ```

Just curious: why not `defer`?

---

*[sql/distsql/server.go, line 48 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7959#-KNc9LGhMZnRs-ju-NSc:-KNc9LGhMZnRs-ju-NSd:-1999355907) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a11041412870fd650d60a25125eb0c48bb02ad1/sql/distsql/server.go#L48)):*
> ```Go
> }
> 
> const flowStreamTimeout time.Duration = 500 * time.Millisecond
> ```

Comment (also that timeout seems too aggressive).

---

*[sql/distsql/server.go, line 94 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7959#-KNc9WGjTfwbYsB3xRK-:-KNc9WGjTfwbYsB3xRK0:-1072540703) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a11041412870fd650d60a25125eb0c48bb02ad1/sql/distsql/server.go#L94)):*
> ```Go
> 	err := f.setupFlow(flowSpec)
> 	if err != nil {
> 		log.Errorc(ds, err.Error(), "", err)
> ```

Enjoy the rebase.

---

*[sql/distsql/server.go, line 108 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7959#-KNc9_gbOp6a9AAAFoc1:-KNc9_gbOp6a9AAAFoc2:-1379581874) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a11041412870fd650d60a25125eb0c48bb02ad1/sql/distsql/server.go#L108)):*
> ```Go
> 	}
> 	// Set up the outgoing mailbox for the stream.
> 	mbox := newOutboxSimpleFlowStream(stream.Context(), stream)
> ```

Needs a similar `context` TODO?

---

*[sql/distsql/tablereader.go, line 215 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7959#-KNc4Selc8NDx--2H_AF:-KNc4Selc8NDx--2H_AG:1462199863) ([raw file](https://github.com/cockroachdb/cockroach/blob/f254bcc6fa60338eb186841acc8ac2f531424a92/sql/distsql/tablereader.go#L215)):*
> ```Go
> 
> 	if log.V(2) {
> 		// TODO(radu): add info about the joinreader in the context.
> ```

Nitto (=nit+ditto).

---

*[sql/distsql/tablereader.go, line 217 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7959#-KNc4YhhrLsf5ji4uN8m:-KNc4YhhrLsf5ji4uN8n:503555143) ([raw file](https://github.com/cockroachdb/cockroach/blob/f254bcc6fa60338eb186841acc8ac2f531424a92/sql/distsql/tablereader.go#L217)):*
> ```Go
> 		// TODO(radu): add info about the joinreader in the context.
> 		log.Infoc(tr.ctx, "TableReader starting (filter: %s)", tr.filter)
> 		defer log.Infoc(tr.ctx, "TableReader exiting")
> ```

Looks like that "TableReader" should just be keyed into the `context.Context` as well.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7959#-:-KNcAAktneyXHHx8oYnL:-1347820123)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
TFTR! Rebased and updated. Current commits r7-r12

---

Review status: 6 of 22 files reviewed at latest revision, 12 unresolved discussions, some commit checks pending.

---

*[sql/distsql/joinreader.go, line 97 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7959#-KNc46yRlmISFMXAYhAR:-KNcZl2uZXwuseFTeSBD:-1630404209) ([raw file](https://github.com/cockroachdb/cockroach/blob/f254bcc6fa60338eb186841acc8ac2f531424a92/sql/distsql/joinreader.go#L97)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> The TODO (or at least its solution) should be in the constructor.
</details>
I was thinking the new context can be a local variable here. We can't modify the contents of flowCtx in the constructor. Granted it should be outside of that `if`, moved.

---

*[sql/distsql/outbox.go, line 172 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7959#-KNc5hrfeeKq8wFCBxQk:-KNc_yUV9IGk5YHV-mjl:1312990214) ([raw file](https://github.com/cockroachdb/cockroach/blob/be5a1469549b4526af1333981045f63df2559e06/sql/distsql/outbox.go#L172)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Just curious: why not `defer`?
</details>
I could put everything here in a defer function but I found that separating the `mainLoop` made it easier to read.

---

*[sql/distsql/server.go, line 94 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7959#-KNc9WGjTfwbYsB3xRK-:-KNccgL1YJpJRNj5XK0f:-1942363498) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a11041412870fd650d60a25125eb0c48bb02ad1/sql/distsql/server.go#L94)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Enjoy the rebase.
</details>
I hate whoever changed all those logs!

---

*[sql/distsql/tablereader.go, line 217 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7959#-KNc4YhhrLsf5ji4uN8m:-KNc_7ywgQxLYctQ-ZcR:466209824) ([raw file](https://github.com/cockroachdb/cockroach/blob/f254bcc6fa60338eb186841acc8ac2f531424a92/sql/distsql/tablereader.go#L217)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Looks like that "TableReader" should just be keyed into the `context.Context` as well.
</details>
Agreed, will do that when I add the contexts (I have most of the change but I will put it in a separate PR).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7959#-:-KNcelbBrz9FUfGzNaBL:1705340646)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
btw, if you want to do me a (small) favor, acknowledge the comments you address (either by `Done.` if they're actionable or with `Acknowledge` if all that counts is that you saw it).

---

Reviewed 16 of 16 files at r7, 7 of 7 files at r8, 2 of 2 files at r9, 2 of 2 files at r10, 4 of 4 files at r11, 2 of 2 files at r12.
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[sql/distsql/joinreader.go, line 97 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7959#-KNc46yRlmISFMXAYhAR:-KNgGp2pStHdaru92b7p:-906123933) ([raw file](https://github.com/cockroachdb/cockroach/blob/f254bcc6fa60338eb186841acc8ac2f531424a92/sql/distsql/joinreader.go#L97)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> I was thinking the new context can be a local variable here. We can't modify the contents of flowCtx in the constructor. Granted it should be outside of that `if`, moved.
</details>
I see, so you have `flowCtx.Context` which is the "user"-supplied context (or some context which is tied to the SQL operation which is going on) and which presumably carries some log tags about the flow as a whole, and then the various subroutines "fork" off of that context instead of you passing that context into any of the moving parts?

---

*[sql/distsql/outbox.go, line 172 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7959#-KNc5hrfeeKq8wFCBxQk:-KNgHIbE2WhdijIuNSLx:594650470) ([raw file](https://github.com/cockroachdb/cockroach/blob/be5a1469549b4526af1333981045f63df2559e06/sql/distsql/outbox.go#L172)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> I could put everything here in a defer function but I found that separating the `mainLoop` made it easier to read.
</details>
No strong feelings, but it always gives me a fuzzy feeling when cleanup is guaranteed to happen. Couldn't you start off with

```go

m.flushTicker = time.NewTicker(outboxFlushPeriod)

defer func() {
  m.flushTicker.Stop()
  // All of the other cleanup stuff
}()

m.err = m.mainLoop()
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7959#-:-KNgIwJTTiL_AgD3WM__:1409991479)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Ah, I always did Ack (and if that comment was a suggestion it means I incorporated it). I'll switch to Done.

---

Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[sql/distsql/joinreader.go, line 97 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7959#-KNc46yRlmISFMXAYhAR:-KNgeTOH92mWV-A_0r8v:1798441257) ([raw file](https://github.com/cockroachdb/cockroach/blob/f254bcc6fa60338eb186841acc8ac2f531424a92/sql/distsql/joinreader.go#L97)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> I see, so you have `flowCtx.Context` which is the "user"-supplied context (or some context which is tied to the SQL operation which is going on) and which presumably carries some log tags about the flow as a whole, and then the various subroutines "fork" off of that context instead of you passing that context into any of the moving parts?
</details>
Correct.

---

*[sql/distsql/outbox.go, line 172 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7959#-KNc5hrfeeKq8wFCBxQk:-KNgeeR5n80SMq2aVrAt:-132926138) ([raw file](https://github.com/cockroachdb/cockroach/blob/be5a1469549b4526af1333981045f63df2559e06/sql/distsql/outbox.go#L172)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> No strong feelings, but it always gives me a fuzzy feeling when cleanup is guaranteed to happen. Couldn't you start off with
> 
> ```go
> 
> m.flushTicker = time.NewTicker(outboxFlushPeriod)
> 
> defer func() {
>   m.flushTicker.Stop()
>   // All of the other cleanup stuff
> }()
> 
> m.err = m.mainLoop()
> ```
</details>
What is the motivation? Are you talking about panics? There is no expectation to recover from panics here (or in most of our code), so I think it's an unnecessary burden to write our code to support that when it doesn't need to.. My main hesitation around using `defer` everywhere is because I don't really know what the performance impact is (of `defer` itself, of the anonymous function, and of any variables which may be moved to heap because of the function). I would be surprised if it's all "free".

Or are you more worried that if cleanup is written the way you suggest, it's less likely the code will be changed and broken in the future?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7959#-:-KNgeBRzcvEPl0mwLndP:-1552152269)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[sql/distsql/outbox.go, line 172 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7959#-KNc5hrfeeKq8wFCBxQk:-KNgi494_OwLmRTHX-VK:-94459507) ([raw file](https://github.com/cockroachdb/cockroach/blob/be5a1469549b4526af1333981045f63df2559e06/sql/distsql/outbox.go#L172)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> What is the motivation? Are you talking about panics? There is no expectation to recover from panics here (or in most of our code), so I think it's an unnecessary burden to write our code to support that when it doesn't need to.. My main hesitation around using `defer` everywhere is because I don't really know what the performance impact is (of `defer` itself, of the anonymous function, and of any variables which may be moved to heap because of the function). I would be surprised if it's all "free".
> 
> Or are you more worried that if cleanup is written the way you suggest, it's less likely the code will be changed and broken in the future?
</details>
I probably would've skipped lightly over this, but I spent a few seconds looking for the `defer` which stops the flush ticker. No real concerns other than that. Between worrying about the performance of `defer` and legibility, I'll always chose legibility unless proven otherwise, but the above code is pretty legible except that it happened to throw me off. I don't think it would've happened had `mainLoop` just taken care of its own ticker, which maybe isn't even used anywhere else.
To sum up, nothing to do here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7959)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Current commits r13-r18.

---

Review status: 6 of 22 files reviewed at latest revision, 9 unresolved discussions, some commit checks pending.

---

*[sql/distsql/outbox.go, line 172 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7959#-KNc5hrfeeKq8wFCBxQk:-KNgjtDbQ6Y8E1pbl9aP:-1858569281) ([raw file](https://github.com/cockroachdb/cockroach/blob/be5a1469549b4526af1333981045f63df2559e06/sql/distsql/outbox.go#L172)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> I probably would've skipped lightly over this, but I spent a few seconds looking for the `defer` which stops the flush ticker. No real concerns other than that. Between worrying about the performance of `defer` and legibility, I'll always chose legibility unless proven otherwise, but the above code is pretty legible except that it happened to throw me off. I don't think it would've happened had `mainLoop` just taken care of its own ticker, which maybe isn't even used anywhere else.
> To sum up, nothing to do here.
</details>
Ah.. Indeed there is no reason `flushTicker` is being set up outside of `mainLoop`, moved it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7959#-:-KNgmlabrdiLiz4CIsXD:1631677630)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

sweet

---

Review status: 6 of 22 files reviewed at latest revision, 22 unresolved discussions, all commit checks successful.

---

*[server/server.go, line 187 \[r14\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7959#-KOBGFv3g_hpRHLOFITx:-KOBGFv40cyFlEDPq-TH:1248381485) ([raw file](https://github.com/cockroachdb/cockroach/blob/ffbaa059c2269e5cfbaea23c9e343b2387c96319/server/server.go#L187)):*
> ```Go
> 	// Set up the DistSQL server
> 	distSQLCtx := distsql.ServerContext{
> 		// TODO(radu): set node ID in the context.
> ```

what's preventing this TODO? :)

---

*[sql/distsql/data.go, line 24 \[r17\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7959#-KOBUiLWbUyVwmFYdkNG:-KOBUiLWbUyVwmFYdkNH:1432730647) ([raw file](https://github.com/cockroachdb/cockroach/blob/6bef55f0e29fa80f12e4eac4437b3084430e3071/sql/distsql/data.go#L24)):*
> ```Go
> )
> 
> func convertColumnOrdering(specOrdering Ordering) sqlbase.ColumnOrdering {
> ```

fine function, but does it really deserve a file?

---

*[sql/distsql/flow.go, line 75 \[r17\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7959#-KOB_BCbgD49LCfn24xF:-KOB_BCbgD49LCfn24xG:116060100) ([raw file](https://github.com/cockroachdb/cockroach/blob/6bef55f0e29fa80f12e4eac4437b3084430e3071/sql/distsql/flow.go#L75)):*
> ```Go
> 
> 	inboundStreams map[StreamID]RowReceiver
> 	localStreams   map[LocalStreamID]*RowChannel
> ```

maybe rename or explain that these localStreams are just here while awaiting the producer to connect.

---

*[sql/distsql/flow.go, line 96 \[r17\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7959#-KOBW-4G2m38-irGoliy:-KOBW-4G2m38-irGoliz:-956331002) ([raw file](https://github.com/cockroachdb/cockroach/blob/6bef55f0e29fa80f12e4eac4437b3084430e3071/sql/distsql/flow.go#L96)):*
> ```Go
> // setupInStream adds a stream to the stream map (inboundStreams or
> // localStreams).
> func (f *Flow) setupInStream(spec StreamEndpointSpec, rowChan *RowChannel) error {
> ```

nit: s/InStream/InputStream or InboundStream

---

*[sql/distsql/flow.go, line 116 \[r17\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7959#-KOBYq_XjZF8OFLCT83_:-KOBYq_XjZF8OFLCT83a:-1653127645) ([raw file](https://github.com/cockroachdb/cockroach/blob/6bef55f0e29fa80f12e4eac4437b3084430e3071/sql/distsql/flow.go#L116)):*
> ```Go
> 	sid := spec.LocalStreamID
> 	if _, found := f.localStreams[sid]; found {
> 		return errors.Errorf("stream %d has multiple connections", sid)
> ```

I got a bit confused about this - maybe say "local stream connected to multiple mailboxes" 

---

*[sql/distsql/flow.go, line 118 \[r17\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7959#-KOBYi5OrKJoTuZi9K8h:-KOBYi5OrKJoTuZi9K8i:-1588164335) ([raw file](https://github.com/cockroachdb/cockroach/blob/6bef55f0e29fa80f12e4eac4437b3084430e3071/sql/distsql/flow.go#L118)):*
> ```Go
> 		return errors.Errorf("stream %d has multiple connections", sid)
> 	}
> 	if f.localStreams[sid] == nil {
> ```

I think you want `f.localStreams == nil`. 
Maybe add a test with multiple local streams?

---

*[sql/distsql/flow.go, line 204 \[r17\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7959#-KOBaNuNXxFpmfiAwgIu:-KOBaNuOXlIJ2T9PxjvY:34402461) ([raw file](https://github.com/cockroachdb/cockroach/blob/6bef55f0e29fa80f12e4eac4437b3084430e3071/sql/distsql/flow.go#L204)):*
> ```Go
> 	// This needs to be done as a separate first stage because unordered input
> 	// synchronizers work by having all sources push into the same RowChannel;
> 	// so the association of RowChannels to streams depends on the input
> ```

I don't quite understand what this comment is trying to say... Maybe rephrase or leave it out.

---

*[sql/distsql/inbound.go, line 28 \[r16\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7959#-KOBSZ_m29niNmkGqiG-:-KOBSZ_m29niNmkGqiG0:741174228) ([raw file](https://github.com/cockroachdb/cockroach/blob/4cf856f56656bfcd5dbef1471b739368cf508d54/sql/distsql/inbound.go#L28)):*
> ```Go
> // them to a RowReceiver. Optionally processes an initial StreamMessage that
> // was already received.
> func ProcessInboundStream(
> ```

mind suggesting in the comment what the deal with the `firstMsg` might be? How come it was read from the stream?

---

*[sql/distsql/inbound.go, line 38 \[r16\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7959#-KOBSJHD4-0P3j2IVbDt:-KOBSJHD4-0P3j2IVbDu:225611922) ([raw file](https://github.com/cockroachdb/cockroach/blob/4cf856f56656bfcd5dbef1471b739368cf508d54/sql/distsql/inbound.go#L38)):*
> ```Go
> 				log.Errorf(ctx, "inbound stream error: %s", err)
> 			}
> 			// TODO(radu): populate response and send error instead
> ```

isn't this important? Otherwise, what happens to `stream`?

---

*[sql/distsql/outbox.go, line 68 \[r15\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7959#-KOBO5-zpSoAdBcygR_4:-KOBO5-zpSoAdBcygR_5:1310883321) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed2d8f6df58431d07305eef786c83db190fdd28a/sql/distsql/outbox.go#L68)):*
> ```Go
> func newOutbox(flowCtx *FlowCtx, addr string, flowID FlowID, streamID StreamID) *outbox {
> 	m := &outbox{flowCtx: flowCtx, addr: addr}
> 	m.encoder.setHeaderFields(flowID, streamID)
> ```

wanna set these fields on the encoder to some reserved values in `newOutboxSimpleFlowStream` too, for symmetry?

---

*[sql/distsql/server.go, line 36 \[r17\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7959#-KOBc-L01JEZqTNkWtEC:-KOBc-L1OjhiDe1lSgXz:-1086568974) ([raw file](https://github.com/cockroachdb/cockroach/blob/6bef55f0e29fa80f12e4eac4437b3084430e3071/sql/distsql/server.go#L36)):*
> ```Go
> // DistSQLServer.
> type ServerContext struct {
> 	context.Context
> ```

Is embedding a context here sane? Should the distsql Server create new top-level contexts for every RPC/flow?

---

*[sql/distsql/server.go, line 133 \[r17\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7959#-KOBfU7AQxE91DSqgxCc:-KOBfU7AQxE91DSqgxCd:244360495) ([raw file](https://github.com/cockroachdb/cockroach/blob/6bef55f0e29fa80f12e4eac4437b3084430e3071/sql/distsql/server.go#L133)):*
> ```Go
> ) {
> 	// For now we don't support multiple flows. We should only need this if we
> 	// want to start flows for multiple queries at the same time.
> ```

how about renaming to `SetupFlow` then?

---

*[sql/distsql/server.go, line 193 \[r17\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7959#-KOBiWtQ2d_THbESsYDk:-KOBiWtRK33a5lVaDU6S:-321624416) ([raw file](https://github.com/cockroachdb/cockroach/blob/6bef55f0e29fa80f12e4eac4437b3084430e3071/sql/distsql/server.go#L193)):*
> ```Go
> // FlowStream is part of the DistSQLServer interface.
> func (ds *ServerImpl) FlowStream(stream DistSQL_FlowStreamServer) error {
> 	err := ds.flowStreamInt(stream)
> ```

do we need the `Int` variant?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7959#-:-KOBjbJN4yrVby46zQ8h:1334692287)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Updated. r19-r25, with the recent changes in r25. Will squash before merging.

---

Review status: 2 of 25 files reviewed at latest revision, 22 unresolved discussions, some commit checks pending.

---

*[server/server.go, line 187 \[r14\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7959#-KOBGFv3g_hpRHLOFITx:-KOBn1wdG_Wgf2IN0hoQ:1058861157) ([raw file](https://github.com/cockroachdb/cockroach/blob/ffbaa059c2269e5cfbaea23c9e343b2387c96319/server/server.go#L187)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> what's preventing this TODO? :)
</details>
I did that in a separate PR #8066 since this was already big.

---

*[sql/distsql/data.go, line 24 \[r17\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7959#-KOBUiLWbUyVwmFYdkNG:-KOBnNJeDwSmhjYl5mDh:-306394788) ([raw file](https://github.com/cockroachdb/cockroach/blob/6bef55f0e29fa80f12e4eac4437b3084430e3071/sql/distsql/data.go#L24)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> fine function, but does it really deserve a file?
</details>
It's associated with data.pb.go.. I expect we'll add more stuff over time

---

*[sql/distsql/flow.go, line 263 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7959#-KNc9-djrkhW9fgY3_rQ:-KOBnwjc9AmikiJ8Cjzw:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a11041412870fd650d60a25125eb0c48bb02ad1/sql/distsql/flow.go#L263)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> nit: s/$/./
</details>
Done.

---

*[sql/distsql/flow.go, line 295 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7959#-KNc98QQq1jjjLmeeSUo:-KOBnwzsMPimdx0RLgAg:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a11041412870fd650d60a25125eb0c48bb02ad1/sql/distsql/flow.go#L295)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Check against double-`Cleanup()`?
</details>
Done.

---

*[sql/distsql/flow.go, line 75 \[r17\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7959#-KOB_BCbgD49LCfn24xF:-KOBrEFAynTxmB794Fb1:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/6bef55f0e29fa80f12e4eac4437b3084430e3071/sql/distsql/flow.go#L75)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> maybe rename or explain that these localStreams are just here while awaiting the producer to connect.
</details>
Done.

---

*[sql/distsql/flow.go, line 96 \[r17\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7959#-KOBW-4G2m38-irGoliy:-KOBrH8fN-IQyn_jRLtE:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/6bef55f0e29fa80f12e4eac4437b3084430e3071/sql/distsql/flow.go#L96)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> nit: s/InStream/InputStream or InboundStream
</details>
Done.

---

*[sql/distsql/flow.go, line 116 \[r17\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7959#-KOBYq_XjZF8OFLCT83_:-KOBrqhbqY26kz00v5LS:1658670501) ([raw file](https://github.com/cockroachdb/cockroach/blob/6bef55f0e29fa80f12e4eac4437b3084430e3071/sql/distsql/flow.go#L116)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> I got a bit confused about this - maybe say "local stream connected to multiple mailboxes" 
</details>
It's kind of the opposite, changed to `local stream %d has multiple consumers`

---

*[sql/distsql/flow.go, line 118 \[r17\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7959#-KOBYi5OrKJoTuZi9K8h:-KOBqigQ6pnGicxuXqtN:422879044) ([raw file](https://github.com/cockroachdb/cockroach/blob/6bef55f0e29fa80f12e4eac4437b3084430e3071/sql/distsql/flow.go#L118)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> I think you want `f.localStreams == nil`. 
> Maybe add a test with multiple local streams?
</details>
Good catch!

---

*[sql/distsql/flow.go, line 204 \[r17\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7959#-KOBaNuNXxFpmfiAwgIu:-KOBsMJMRPRo9wTkh4Yu:1740281843) ([raw file](https://github.com/cockroachdb/cockroach/blob/6bef55f0e29fa80f12e4eac4437b3084430e3071/sql/distsql/flow.go#L204)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> I don't quite understand what this comment is trying to say... Maybe rephrase or leave it out.
</details>
Removed.

---

*[sql/distsql/inbound.go, line 60 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7959#-KNc6h1CQJgCOlo371bv:-KOBnydc62c_VSGzBUfc:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/2dc8311c371f3fe95bf18fd07e75e56dfdc6bf50/sql/distsql/inbound.go#L60)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Instead of all of these `retErr = err; break loop`, could you `return responseError(err)`  which is roughly
> 
> ```go
> responseError := func(err error) error {
>   dst.Close(err)
>   resp := &SimpleResponse{}
>   resp.Error = something(err)
>   return stream.SendAndClose(resp)
> }
> ```
> 
> which would obviate I think all of the `break loop`?
</details>
Done.

---

*[sql/distsql/inbound.go, line 28 \[r16\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7959#-KOBSZ_m29niNmkGqiG-:-KOBo-ntb5orZwNhbeUU:782752367) ([raw file](https://github.com/cockroachdb/cockroach/blob/4cf856f56656bfcd5dbef1471b739368cf508d54/sql/distsql/inbound.go#L28)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> mind suggesting in the comment what the deal with the `firstMsg` might be? How come it was read from the stream?
</details>
The first message contains the info about what flow we are trying to connect to.. So we have to read that first message before we can get here. I'll add to the comment.

---

*[sql/distsql/inbound.go, line 38 \[r16\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7959#-KOBSJHD4-0P3j2IVbDt:-KOBoATLZjGRcyqQBhDH:1497353687) ([raw file](https://github.com/cockroachdb/cockroach/blob/4cf856f56656bfcd5dbef1471b739368cf508d54/sql/distsql/inbound.go#L38)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> isn't this important? Otherwise, what happens to `stream`?
</details>
I believe the caller would receive the error (similar to normal RPCs).

---

*[sql/distsql/outbox.go, line 68 \[r15\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7959#-KOBO5-zpSoAdBcygR_4:-KOBtChdpbpMdiiD4f8_:12022312) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed2d8f6df58431d07305eef786c83db190fdd28a/sql/distsql/outbox.go#L68)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> wanna set these fields on the encoder to some reserved values in `newOutboxSimpleFlowStream` too, for symmetry?
</details>
the 0 values are just fine..

---

*[sql/distsql/server.go, line 36 \[r17\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7959#-KOBc-L01JEZqTNkWtEC:-KOBoWlGCW2xln2iZxK7:714374867) ([raw file](https://github.com/cockroachdb/cockroach/blob/6bef55f0e29fa80f12e4eac4437b3084430e3071/sql/distsql/server.go#L36)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> Is embedding a context here sane? Should the distsql Server create new top-level contexts for every RPC/flow?
</details>
Huh? It's created once for the entire server. I'm embedding just to be able to pass to `log.Mehf` directly but I can make it a member (it's the same thing otherwise).

---

*[sql/distsql/server.go, line 133 \[r17\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7959#-KOBfU7AQxE91DSqgxCc:-KOBtH5ztjh46Dwuegyr:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/6bef55f0e29fa80f12e4eac4437b3084430e3071/sql/distsql/server.go#L133)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> how about renaming to `SetupFlow` then?
</details>
Done.

---

*[sql/distsql/server.go, line 193 \[r17\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7959#-KOBiWtQ2d_THbESsYDk:-KOBotWrMImn3PGb28y-:1672824606) ([raw file](https://github.com/cockroachdb/cockroach/blob/6bef55f0e29fa80f12e4eac4437b3084430e3071/sql/distsql/server.go#L193)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> do we need the `Int` variant?
</details>
just so it can do `return` without logging everywhere.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7959#-:-KOBw5zi1Lpiu86k3zUn:-2037992460)*
<!-- Sent from Reviewable.io -->

RaduBerinde committed:
distsql: streams infrastructure and multi-node test

Support for flow-local and inter-node streams:
 - flow set up logic for streams
 - inbound stream handling
 - outbox logic to connect to other hosts using FlowStream
 - 3-node distributed flow test

Part of #7583.

RaduBerinde commented:
Squashed into a single commit.

---

Review status: 2 of 24 files reviewed at latest revision, 22 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7959#-:-KOCLaEizdg-t3VBmBYm:-1466549439)*
<!-- Sent from Reviewable.io -->
